### PR TITLE
Add option to use update instead of patch in graph upserts

### PR DESCRIPTION
**Pull request by lehni**
As discussed [before](https://github.com/Vincit/objection.js/issues/557#issuecomment-336901820), here now the PR to add this.

This is a follow up PR for #583, I will rebase once that one is merged.

Please note that there are things to be figured out still, as outlined in this comment:

```js
// TODO: Should we distinguish between update and patch for noUpdate also?
// If we want different options, then:
// return hasOption(node, 'update')
//   ? decideType(node, UpsertNodeType.Update, 'noUpdate')
//   : decideType(node, UpsertNodeType.Patch, 'noPatch');
```

Note that `'noUpdate'` in the current code base would become `'noPatch'` if we did this.

### CODE REVIEW

File: lib/queryBuilder/graphUpserter/UpsertNode.js [hunk 1]
```diff
@@ -131,15 +110,18 @@ function getType(node) {
...
-        return UpsertNodeType.Update;
-      }
+      return decideType(node,
+        hasOption(node, 'update') ? UpsertNodeType.Update : UpsertNodeType.Patch,
+        'noUpdate')
+      // TODO: Should we distinguish between update and patch for noUpdate also?
+      // If we want different options, then:
+      // return hasOption(node, 'update')
+      //   ? decideType(node, UpsertNodeType.Update, 'noUpdate')
+      //   : decideType(node, UpsertNodeType.Patch, 'noPatch');
```
Comment (koskimas) [inline lib/queryBuilder/graphUpserter/UpsertNode.js]: I don't think there is any need for separate `noUpdate` and `noPatch`.

**Comment by koskimas** [comment]
This still needs tests and documentation.

**Comment by lehni** [comment]
@koskimas yes I was waiting with this until the open question was decided. I just rebased and merged with #527, will work on tests this week.

**Comment by lehni** [comment]
@koskimas I can't figure out how to test if patch is used instead of update when passing the option... It looks like the only difference is really in validation (through `modelOptions` that get passed on), so I guess I could cause a validation error that only happens when doing an incomplete update instead of a patch. But how to achieve this in `tests/integration/upsertGraph.js`?

**Comment by lehni** [comment]
I've rebased again now and merged all conflicts. I hope you're not planning another general large-scale change of code conventions any time soon, @koskimas :)

**Comment by koskimas** [comment]
Sorry about that @lehni. I merged all your PR's that I was able before running `prettier`.

To answer your earlier question, you can update a model's jsonSchema like [this](https://github.com/Vincit/objection.js/blob/master/tests/integration/insertGraph.js#L122). It's a bit hacky, but gets the job done.

**Comment by lehni** [comment]
@koskimas I've tried to achieve this test for the past hour, but failed... Could you please take a look? I've pushed it here now. From my own tests separately, a `upserGraph(..., { update: true })` does fail validation if the provided data is not complete. I don't understand why in the test this isn't happening, but maybe it has something to do with internal caching and the fact that another test with another `jsonSchema` runs right before it?

File: tests/integration/upsertGraph.js [hunk 2]
```diff
@@ -1777,6 +1778,110 @@ module.exports = session => {
...
+        delete Model1.$$jsonSchema;
+        delete Model1.$$validator;
+      });
+
+      it('should fail to do an incomplete upsert', () => {
+        const fails = [
+          {
+            id: 2,
+            // This fails because of missing property.
+            // model1Prop1: 'updated root 2',
```
Comment (koskimas) [inline tests/integration/upsertGraph.js]: There's an optimization that no updates are done if there is only an identifier present. Your code never reaches the validation line. A better way to test this would be to set the `model1Prop1` and give it a value of wrong type.

File: tests/integration/upsertGraph.js [hunk 3]
```diff
@@ -1777,6 +1778,110 @@ module.exports = session => {
...
+              model1Prop1: 'updated belongsToOne'
+            }
+          }
+        ];
+
+        return Promise.map(fails, fail => {
+          return transaction(session.knex, trx =>
+            Model1.query(trx).upsertGraph(fail, {update: true})
+          ).reflect();
+        })
```
Comment (koskimas) [inline tests/integration/upsertGraph.js]: Sigh... I hate the way prettier aligns code with these. Well, there nothing we can do..

Comment (lehni) [reply tests/integration/upsertGraph.js]: I guess you can configure it?

File: tests/integration/upsertGraph.js (hunk 2 above)
Comment (lehni) [reply tests/integration/upsertGraph.js]: Oh I see, that makes sense. But I need to catch the case where an update does not provide all required properties and hence causes that kind of validation error, which a patch wouldn't cause due to the use of `skipRequired` in compileValidator() (which as far as I can tell is the only difference between patch and update?)

File: tests/integration/upsertGraph.js (hunk 3 above)
Comment (koskimas) [reply tests/integration/upsertGraph.js]: Nope, the there is not much to configure in prettier.

File: tests/integration/upsertGraph.js (hunk 2 above)
Comment (koskimas) [reply tests/integration/upsertGraph.js]: The optimization is in `UpsertNode.hasChanges` it tests if any of the values has actually changed. Since we only fetch ids and foreign keys from the db, the test only compares those. I think the way it currently works is correct. If there is nothing to update, even if the model is invalid, nothing should be done.

Comment (koskimas) [reply tests/integration/upsertGraph.js]: We can't move the validation to an earlier position, since some properties get set by the insert part. The validation must be done after insert, but before update.

Comment (lehni) [reply tests/integration/upsertGraph.js]: Not saying it's not correct. I guess I'd need a model with more than one property to create an incomplete update?

**Comment by lehni** [comment]
@koskimas I believe I've figured it out now: a8af40cd9405d0089761c13b9b505186b4e3fed5

Using two props isntead of one like in the test right before it allows me to produce and test the described scenarios

File: tests/integration/upsertGraph.js (hunk 3 above)
Comment (lehni) [reply tests/integration/upsertGraph.js]: That's why I use a heavily configured ESLint with precommit hooks instead of prettier. Prettier is great, but for me there are too many edge cases where it behaves a tad too weirdly.

**Comment by lehni** [comment]
@koskimas ooops the tests were still failing because deleting `$$jsonSchema` doesn't get rid of the cached `$$validator`. I have to delete both after the first test that overrides `$$jsonSchema` and validates. I've added it in both places now, but you may want to look into caching, since changin `$$jsonSchema` should cause the cache to invalidate, perhaps?

File: tests/integration/upsertGraph.js [hunk 4]
```diff
@@ -1551,6 +1551,7 @@ module.exports = session => {
 
       after(() => {
         delete Model1.$$jsonSchema;
+        delete Model1.$$validator;
```
Comment (lehni) [inline tests/integration/upsertGraph.js]: @koskimas this is the addition that was necessary for the following test to get a new validator.

**Comment by lehni** [comment]
Oh wow, now it's failing on postgres. That test doesn't work for me locally, not sure how to get it set up?

```
integration tests postgres upsertGraph upserts with update: true option should fail to do an incomplete upsert:
     error: update "Model1" set "id" = $1, "model1Id" = $2, "model1Prop1" = $3, "model1Prop2" = $4 where "Model1"."id" = $5 - invalid input syntax for integer: "updated root 2, prop 2"
```

**Comment by lehni** [comment]
Oh I see, that property is an integer, not  a string there. That's easy, I'll change the test accordingly.

**Comment by lehni** [comment]
@koskimas ok, I'm pretty confident that this one will turn green... 10th time lucky, or something? Talk of a tiny feature, with a unit test that blows out of proportion!

File: tests/integration/upsertGraph.js (hunk 4 above)
Comment (koskimas) [reply tests/integration/upsertGraph.js]: Oh crap! I think I've made the same mistake in other tests too. I'll check them out. Good catch!

**Comment by koskimas** [comment]
Looking good! Could you add a short description of the `update` option in `UpsertGraphOptions` object's documentation in API.md? Then I'm ready hit merge.

**Comment by lehni** [comment]
Will take care of it tomorrow!

**Comment by lehni** [comment]
@koskimas I've pushed the documentation. PTAL

**Comment by koskimas** [comment]
Awesome++ Thank you so much for fighting with this!

**Comment by lehni** [comment]
The feeling's mutual! : )
